src/storage/btree_load.c

```diff
@@ -3878,3 +3750,457 @@ btree_get_perf_btree_page_type (THREAD_ENTRY * thread_p, PAGE_PTR page_ptr)
+
+    }
+end:
+
+  if (old_page != NULL)
+    {
+      pgbuf_unfix_and_init (thread_p, old_page);
+    }
+
+  if (curr_fk_pageptr != NULL)
+    {
+      pgbuf_unfix_and_init (thread_p, curr_fk_pageptr);
+    }
+
+  if (pk_bt_scan.C_page)
+    {
+      pgbuf_unfix_and_init (thread_p, pk_bt_scan.C_page);
+    }
+
+  if (load_args->leaf.pgptr)
+    {
+      pgbuf_unfix (thread_p, load_args->leaf.pgptr);
+    }
+
+  if (!DB_IS_NULL (&fk_key))
+    {
+      pr_clear_value (&fk_key);
+    }
+  if (!DB_IS_NULL (&pk_key))
+    {
+      pr_clear_value (&pk_key);
+    }
+
+  return ret;
+}
+
+/*
+ * btree_get_value_from_leaf_slot () -
+ *
+ *   return:			  - NO_ERROR or error code.
+ *   btid_int(in):		  - The structure of the B-tree where the leaf resides.
+ *   leaf_ptr(in):		  - The leaf where the value needs to be extracted from.
+ *   slot_id(in):		  - The slot from where the value must be pulled.
+ *   key(out):			  - The value requested.
+ *
+ */
+static int
+btree_get_value_from_leaf_slot (THREAD_ENTRY * thread_p, BTID_INT * btid_int, PAGE_PTR leaf_ptr, int slot_id,
+				DB_VALUE * key)
+{
+  LEAF_REC leaf;
+  bool clear_first_key = false;
+  int first_key_offset = 0;
+  RECDES record;
+  int ret = NO_ERROR;
+
+  if (spage_get_record (thread_p, leaf_ptr, slot_id, &record, PEEK) != S_SUCCESS)
+    {
+      assert_release (false);
+      ret = ER_FAILED;
+      return ret;
+    }
+
+  ret =
+    btree_read_record (thread_p, btid_int, leaf_ptr, &record, key,
+		       &leaf, BTREE_LEAF_NODE, &clear_first_key, &first_key_offset, PEEK_KEY_VALUE, NULL);
+
+  if (ret != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return ret;
+    }
+
+  return ret;
+}
+
+/*
+ * btree_advance_to_next_slot_and_fix_page () -
+ *
+ *   return:			  - NO_ERROR or error code
+ *   btid(in):			  - B-tree structure
+ *   vpid(in/out):		  - VPID of the current page
+ *   pg_ptr(in/out):		  - Page pointer for the current page.
+ *   slot_id(in/out):		  - Slot id of the current/next value.
+ *   key(out):			  - Requested key.
+ *   key_cnt(in/out):		  - Number of keys in current page.
+ *   header(in/out):		  - The header of the current page.
+ *
+ *  Note:
+ *    - slot_id will point towards the current slot from where the value
+ *	will be extracted at the beginning of the function;
+ *	on exit, it will point towards the next value that needs to be extracted.
+ */
+static int
+btree_advance_to_next_slot_and_fix_page (THREAD_ENTRY * thread_p, BTID_INT * btid, VPID * vpid, PAGE_PTR * pg_ptr,
+					 INT16 * slot_id, DB_VALUE * key, bool is_desc, int *key_cnt,
+					 BTREE_NODE_HEADER ** header, MVCC_SNAPSHOT * mvcc)
+{
+  int ret = NO_ERROR;
+  VPID next_vpid;
+  PAGE_PTR page = *pg_ptr;
+  BTREE_NODE_HEADER *local_header = *header;
+  int adv = 0, num_visible = -1;
+  bool first_scan = false;
+  PAGE_PTR old_page = NULL;
+
+  /* Clear current key, if any. */
+  if (!DB_IS_NULL (key))
+    {
+      pr_clear_value (key);
+    }
+
+start_alg:
+  if (page != NULL)
```

Comment:
The function should handle page change. Also, it's better to avoid goto labels.

do something like:

```

if not initialized (current page is nil)
  current page  = fix page
  check error
  initialize slot, node key count and node header

/* advance to next key (having at least one visible object) */
while (true) {
  increment slot;
  if slot not in page
    if no next/prev page
      unfix current page
      break loop
    else
      next page = fix next/prev page
      check error
      unfix current page
      current page = next page
      initialize slot, node key count and node header
  if snapshot != NULL
    count visible objects
    check error
    if visible == 0
      continue
  /* key is valid */
  break
}
```